rclcpp/minimal_action_client/member_function.cpp

```diff
@@ -0,0 +1,123 @@
+  using GoalHandleFibonacci = rclcpp_action::ClientGoalHandle<Fibonacci>;
+
+  MinimalActionClient(const std::string node_name)
+    : Node(node_name)
+  {
+  }
+
+  void spin_until_goal_complete(const int goal_order)
+  {
+    using namespace std::placeholders;
+
+    auto client_ptr =
+      rclcpp_action::create_client<Fibonacci>(this->shared_from_this(), "fibonacci");
+
+    if (!client_ptr->wait_for_action_server(std::chrono::seconds(20))) {
+      RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
+      return;
+    }
+
+    auto goal_msg = Fibonacci::Goal();
+    goal_msg.order = goal_order;
+
+    RCLCPP_INFO(this->get_logger(), "Sending goal");
+    auto goal_handle_future = client_ptr->async_send_goal(
+      goal_msg,
+      std::bind(&MinimalActionClient::feedback_callback, this, _1, _2));
+
+    if (rclcpp::spin_until_future_complete(this->shared_from_this(), goal_handle_future) !=
+      rclcpp::executor::FutureReturnCode::SUCCESS)
+    {
+      RCLCPP_ERROR(this->get_logger(), "Send goal call failed :(");
+      return;
+    }
+
+    GoalHandleFibonacci::SharedPtr goal_handle = goal_handle_future.get();
+    if (!goal_handle) {
+      RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
+      return;
+    }
+
+    // Wait for the server to be done with the goal
+    auto result_future = goal_handle->async_result();
+
+    RCLCPP_INFO(this->get_logger(), "Waiting for result");
+    if (rclcpp::spin_until_future_complete(this->shared_from_this(), result_future) !=
+      rclcpp::executor::FutureReturnCode::SUCCESS)
+    {
+      RCLCPP_ERROR(this->get_logger(), "Get result call failed :(");
+      return;
+    }
+
+    GoalHandleFibonacci::WrappedResult wrapped_result = result_future.get();
+
+    switch(wrapped_result.code) {
+      case rclcpp_action::ResultCode::SUCCEEDED:
+        break;
+      case rclcpp_action::ResultCode::ABORTED:
+        RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
+        return;
+      case rclcpp_action::ResultCode::CANCELED:
+        RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
+        return;
+      default:
+        RCLCPP_ERROR(this->get_logger(), "Unknown result code");
+        return;
+    }
+
+    RCLCPP_INFO(this->get_logger(), "Result received");
+    for (auto number : wrapped_result.result->sequence)
+    {
+      RCLCPP_INFO(this->get_logger(), "%" PRId64, number);
+    }
+  }
+
+private:
+  void feedback_callback(
+    GoalHandleFibonacci::SharedPtr,
+    const std::shared_ptr<const Fibonacci::Feedback> feedback)
+  {
+    RCLCPP_INFO(
+      this->get_logger(),
+      "Next number in sequence received: %" PRId64,
+      feedback->sequence.back());
+  }
+};  // class MinimalActionClient
+
+int main(int argc, char ** argv)
+{
+  rclcpp::init(argc, argv);
+  auto action_client = std::make_shared<MinimalActionClient>("minimal_action_client");
+
+  action_client->spin_until_goal_complete(10);
```

Comment:
This works, but it would be difficult to compose because it needs to call the executor to block. It would be easier if there was to run a function when a future completes.

Reply:
I guess we should add callbacks as described in https://github.com/ros2/rclcpp/issues/635 to make this example better.

Reply:
I've refactored the composable example to use callbacks in 82c7acb. This depends on ros2/rclcpp#701.